### ISSUE TITLE
HW 2 review by OttoS

### ISSUE THREAD

**Issue by ostegm** [iswearingen/DAT_SF_12#1]
@iswearingen

Hi India, 

The "official" solutions are posted now. You should be able to do a git pull and find those for review. They're also located here: https://github.com/ga-students/DAT_SF_12/tree/gh-pages/Solutions

So... some comments, in chronological order running through your homework. 
1. Your application of K-means and cross-validation is great. You arrived at the correct answer, however I wanted to point out a few subtle points. It looks like you used the code from class (which is awesome) - but be sure to think about what the inputs are. For example, you looped through the list of odd integers from 1 to 51.

``` python
n_neighbors = range(1, 51, 2)
```

Why 51? That was the number of observations we had in the demo example. In this case 51 isn't a bad choice, but I wanted to make sure it was conscious. In this data set we have 178 observations, so technically you could go to 177, although I think 51 is a better choice.... Anyway, small point, but wanted to make sure you understood why we used 51. 
1. When you chose the number of neighbors (30) I see that you based this off of the graph. Which optimized the score at 27. Remember that this graph was built off of only one slice of the data: 

If you were to run the code below (with a random seed of 1, you'd find a different value for K. 

``` python
X_train, X_test, y_train, y_test = train_test_split(wine_variables, wine['Wine'], test_size=0.3, random_state=1)
```

The point is, be careful picking K values based off of a random slice of data. Sometimes you can end up overfitting the model based on a single slice of the data. Another way to do this would be to fit the model, and the score based on cross-validation before choosing your K value. 

Additionally, you usually want to use a odd value for K so that there are no tie's (if I use 2 nearest neighbors and both are different classes, which way to I go)
1. This is kind of bonus material, but did you notice that proline is of a bigger magnitude than all the other variables? It ranges from 0-168 when some of the other variables a much much smaller. This scale problem over amplifies the effect of proline. If you scale the data. you can get an accuracy of 96%.

``` python
from sklearn.preprocessing import StandardScaler
features_scalar = StandardScaler()
X_train_scaled = features_scalar.fit_transform(X_train)

from sklearn import neighbors
clf_scaled = neighbors.KNeighborsClassifier(3, weights='uniform')
clf_scaled.fit(X_train_scaled, y_train)
```
1. In the clustering section, I like that you did the pd.crosstab method to see which classes weren't lining up. Great! Just remember that the class labels won't match up. A zero won't necessarily be a zero in other predictions since they are just arbitrary numbers to the clustering algorithm... But you can still see the patterns. 

Let me know if you have questions! 

Thanks

**Comment by iswearingen** [comment]
Thanks for the feedback. I desperately want to understand all of the tips you gave me, so had some follow-up things to ask if you have time: 

In your review you mentioned picking the right range, I wasn't quit sure what this meant and how to do it. Although you mentioned 51 was good for this example, how do you choose what the right number is? Doesnt sound like its always a rule of thumb to use the entire observation. 

In the random seed comment, I am not sure I understand this part either. I get the general concept of overfitting, but I dont know what you mean by testing out different slices of the data, any tools, tips, things that can dumb the concept down a little more?

And lastly, just to clarify the scaling comment, are you simply saying that if I were to scale the data, I would have seen a much better fit than the 70 something percent I got?

Thanks so much for reviewing!

**Comment by ostegm** [comment]
Maybe we can chat about this in person on Thursday?

**Comment by iswearingen** [comment]
Yes please! What time will you be in?

**Comment by ostegm** [comment]
I can try to get their by 6:15, that work?

**Comment by iswearingen** [comment]
Yes! I will be there. Thank you so much.

On Wed, Feb 25, 2015 at 9:12 AM, Otto Stegmaier notifications@github.com
wrote:

> I can try to get their by 6:15, that work?
> 
> On Wed, Feb 25, 2015 at 8:34 AM, iswearingen notifications@github.com
> wrote:
> 
> > Yes please! What time will you be in?
> > 
> > ## 
> > 
> > Reply to this email directly or view it on GitHub
> > <https://github.com/iswearingen/DAT_SF_12/issues/1#issuecomment-75993958
> > .
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/iswearingen/DAT_SF_12/issues/1#issuecomment-76005392.

## 

India Swearingen
MA, Applied Social Psychology and Program Evaluation
India.Swearingen@gmail.com
(909)645-6356

**Comment by ostegm** [comment]
Running a few behind be there in 5!
